programs/drift/src/state/fulfillment_params/phoenix.rs

```diff
@@ -0,0 +1,479 @@
+            phoenix_base_vault,
+            phoenix_quote_vault,
+            token_program,
+            base_market_vault,
+            quote_market_vault,
+            signer_nonce: state.signer_nonce,
+        })
+    }
+}
+
+impl<'a, 'b> PhoenixFulfillmentParams<'a, 'b> {
+    pub fn to_account_infos(&self) -> [AccountInfo<'b>; 9] {
+        [
+            self.phoenix_program.clone(),
+            self.phoenix_log_authority.clone(),
+            self.phoenix_market.to_account_info(),
+            self.drift_signer.clone(),
+            self.base_market_vault.to_account_info(),
+            self.quote_market_vault.to_account_info(),
+            self.phoenix_base_vault.to_account_info(),
+            self.phoenix_quote_vault.to_account_info(),
+            self.token_program.to_account_info(),
+        ]
+    }
+}
+
+impl<'a, 'b> PhoenixFulfillmentParams<'a, 'b> {
+    pub fn invoke_new_order(&self, order_packet: OrderPacket) -> DriftResult {
+        let base_mint = self.phoenix_market.header.base_params.mint_key;
+        let quote_mint = self.phoenix_market.header.quote_params.mint_key;
+
+        let new_order_instruction = create_new_order_instruction_with_custom_token_accounts(
+            self.phoenix_market.key,
+            self.drift_signer.key,
+            &self.base_market_vault.key(),
+            &self.quote_market_vault.key(),
+            &base_mint,
+            &quote_mint,
+            &order_packet,
+        );
+
+        let signer_seeds = get_signer_seeds(&self.signer_nonce);
+        let signers_seeds = &[&signer_seeds[..]];
+
+        invoke_signed_unchecked(
+            &new_order_instruction,
+            &self.to_account_infos(),
+            signers_seeds,
+        )
+        .map_err(|e| {
+            msg!("{:?}", e);
+            ErrorCode::FailedPhoenixCPI
+        })?;
+
+        Ok(())
+    }
+}
+
+impl<'a, 'b> SpotFulfillmentParams for PhoenixFulfillmentParams<'a, 'b> {
+    fn is_external(&self) -> bool {
+        true
+    }
+
+    fn get_best_bid_and_ask(&self) -> DriftResult<(Option<u64>, Option<u64>)> {
+        // Conversion: price_in_ticks (T) * tick_size (QL/BU * T) * quote_lot_size (QA/QL) / raw_base_units_per_base_unit (rBU/BU)
+        // Yields: price (QA/rBU)
+        let market_data = self.phoenix_market.data.borrow();
+        let (_, market_bytes) = market_data.split_at(size_of::<MarketHeader>());
+        let header = &self.phoenix_market.header;
+        if header.quote_params.decimals != 6 {
+            msg!("Quote decimals must be 6");
+            return Err(ErrorCode::InvalidPricePrecision);
+        }
+
+        let market = load_with_dispatch(&header.market_size_params, market_bytes)
+            .map_err(|_| {
+                msg!("Failed to deserialize market");
+                ErrorCode::FailedToDeserializePhoenixMarket
+            })?
+            .inner;
+
+        let best_bid = market.get_book(Side::Bid).iter().next().and_then(|(o, _)| {
```

Comment:
are these precision 1e6? would be nice to have a cargo test for calculating the price, could maybe split it into its own function

Reply:
When the dust settles, these will have the same precision as the quote token. Units are really tricky to get right because "price" isn't really a unit, it's actually a ratio. What you're really looking for here is the number of USDC sats (we call them atoms) per UI unit of base token like SOL (we call this a "raw base unit").

So if you read my comment at the top, our unit conversions should result in correct final price ratio (quote atoms per raw base unit) which has precision 1e6

```
// Conversion: price_in_ticks (T) * tick_size (QL/BU * T) * quote_lot_size (QA/QL) / raw_base_units_per_base_unit (rBU/BU)
// Yields: price (QA/rBU)
```